### ISSUE TITLE
[FasterRCNN] Understanding clip_boxes in generate_fpn_proposals

### ISSUE THREAD

**Issue by armandmcqueen** [tensorpack/tensorpack#1066]
I'm working with the Mask/Faster RCNN code and I'm confused about why the ROI proposals are being generated on a per-FPN-level basis, but the call to `generate_rpn_proposals` for each level passes in the same `image_shape2d`.

In the RPN code, you call `generate_fpn_proposals`. https://github.com/tensorpack/tensorpack/blob/30ead05beea3a8cc4ea3c9af7872a43aa7e9491c/examples/FasterRCNN/train.py#L248-L249

In `generate_fpn_proposals`, you call `generate_rpn_proposals` for each level. https://github.com/tensorpack/tensorpack/blob/30ead05beea3a8cc4ea3c9af7872a43aa7e9491c/examples/FasterRCNN/model_fpn.py#L186-L193

`image_shape2d` is the shape (h,w,) of the 'image' input tensor. In `generate_rpn_proposals`, `image_shape2d` is used in the call to `clip_boxes`, which clips each proposal so all corners are between coordinates (0,0) and (h,w).

https://github.com/tensorpack/tensorpack/blob/30ead05beea3a8cc4ea3c9af7872a43aa7e9491c/examples/FasterRCNN/model_rpn.py#L129

Won't the proposals be using the coordinate system of the feature map, which will be much smaller than the original size of the image? In that case, won't `clip_boxes` only work correctly on corners that exist in the negative coordinate space?

To give an example, if there is an image that is 1024x1024, the backbone gives you layers with sizes 256x256, 128x128, 64x64, etc. 

When you run the RPN head over the 256x256 feature map, if you were to get a proposal with (x1, y1), (x2, y2) = (-10, -20), (200, 300), you want `clip_boxes` to convert that to (0, 0), (200, 256) so that it exists fully within the feature map of this layer. However, it looks like `clip_boxes` will only ensure that the corners are between (0,0) and (1024, 1024) so you get (0,0), (200, 300).

Am I misunderstanding something here?

**Comment by ppwwyyxx** [comment]
> Won't the proposals be using the coordinate system of the feature map, 

No. The proposals are generated from the feature map but their values are in the coordinate system of the input image.

**Comment by armandmcqueen** [comment]
Okay, that's helpful, thank you. And am I correct that: 
- The `rpn_head` outputs `box_logits` which are in the coordinate system of the given FPN layer feature map. 
- Then `anchor.decode_logits(box_logits)` uses the anchors for that layer to convert from `box_logits` to the coordinate system of the input image. 
- This works because each FPN layer is associated with a given anchor size so as you move up in the FPN, the feature space is smaller, but the anchor boxes are larger. 
- The `rpn_head` isn't aware of which FPN layer the input comes from, but because the anchor boxes are larger in higher layers, the same (x1, y1, h1, w1) output of `rpn_head` will lead to a larger proposal when output on FPN_LVL_4 than when output on FPN_LVL_3.

**Comment by ppwwyyxx** [comment]
1. No. Box logits are not defined on any particular coordinates. It only encodes the ratio w.r.t anchors and does not have its own absolute scale.
Refer to the fast rcnn paper for details.

2. Yes. Anchors are defined on coordidates of the input image.

4. See 1.

**Comment by armandmcqueen** [comment]
Ah, I was misunderstanding how `box_logits` were parameterized. That's much clearer now - thank you for the help!
